src/libraries/System.Console/src/System/Console.cs

```diff
@@ -33,31 +33,51 @@ public static class Console
         private static ConsoleCancelEventHandler? s_cancelCallbacks;
         private static ConsolePal.ControlCHandlerRegistrar? s_registrar;
 
-        internal static T EnsureInitialized<T>([NotNull] ref T? field, Func<T> initializer) where T : class =>
-            LazyInitializer.EnsureInitialized(ref field, ref InternalSyncObject, initializer);
+        public static TextReader In
+        {
+            get
+            {
+                return Volatile.Read(ref s_in) ?? EnsureInitializedIn();
 
-        public static TextReader In => EnsureInitialized(ref s_in, () => ConsolePal.GetOrCreateReader());
+                static TextReader EnsureInitializedIn()
+                {
+                    ConsolePal.EnsureConsoleInitialized(); // May lock Console.Out
+
+                    lock (InternalSyncObject) // Ensures In and InputEncoding are synchronized.
+                    {
+                        return s_in ??= ConsolePal.GetOrCreateReader();
+                    }
+                }
+            }
+        }
 
         public static Encoding InputEncoding
         {
             get
             {
-                return EnsureInitialized(ref s_inputEncoding, () => ConsolePal.InputEncoding);
+                Encoding? encoding = Volatile.Read(ref s_inputEncoding);
+                if (encoding == null)
+                {
+                    encoding = ConsolePal.InputEncoding;
+                    Volatile.Write(ref s_inputEncoding, encoding);
```

Comment:
I think this should be done under the same lock as the setter. Otherwise, we have the risk of things getting out of sync again.